### PR TITLE
#3119 Add qualifiedBy and qualifiedByName to SubclassMapping annotation

### LINKED ISSUE #3119: Support qualifier for SubclassMapping annotated methods
### Use case

I want to specify which method the subclass mapping should use. For example sometimes i want to use a more complete method that maps all the fields of a subclass and sometimes i want to use a method that just map some fields. This is currently not possible because the qualifiers specified in the BeanMapping annotations is ignored.

These are my mappers:

```java
@Mapper(uses = Dummy1Mapper.class)
public interface DummyMapper {
    @SubclassMapping(source = Dummy1.class, target = Dummy1Dto.class)
    BaseDummyDto toDto(BaseDummy baseDummy);

    @BeanMapping(qualifiedByName = "light")
    @SubclassMapping(source = Dummy1.class, target = Dummy1Dto.class)
    BaseDummyDto toDtoLight(BaseDummy baseDummy);
}

@Mapper
public interface Dummy1Mapper {
    Dummy1Dto toDto(Dummy1 dummy1);

    @Named("light")
    @Mapping(target = "descriptions", ignore = true)
    Dummy1Dto toDtoLight(Dummy1 dummy1);
}
```

### Generated Code

These are the generated mappers, as you can see Dummy1Mapper#toDtoLight is never called:

```java
public class DummyMapperImpl implements DummyMapper {

    private final Dummy1Mapper dummy1Mapper = Mappers.getMapper( Dummy1Mapper.class );

    @Override
    public BaseDummyDto toDto(BaseDummy baseDummy) {
        if ( baseDummy == null ) {
            return null;
        }

        if (baseDummy instanceof Dummy1) {
            return dummy1Mapper.toDto( (Dummy1) baseDummy );
        }
        else {
            BaseDummyDto baseDummyDto = new BaseDummyDto();

            baseDummyDto.setId( baseDummy.getId() );

            return baseDummyDto;
        }
    }

    @Override
    public BaseDummyDto toDtoLight(BaseDummy baseDummy) {
        if ( baseDummy == null ) {
            return null;
        }

        if (baseDummy instanceof Dummy1) {
            return dummy1Mapper.toDto( (Dummy1) baseDummy );
        }
        else {
            BaseDummyDto baseDummyDto = new BaseDummyDto();

            baseDummyDto.setId( baseDummy.getId() );

            return baseDummyDto;
        }
    }
}

public class Dummy1MapperImpl implements Dummy1Mapper {

    @Override
    public Dummy1Dto toDto(Dummy1 dummy1) {
        if ( dummy1 == null ) {
            return null;
        }

        Dummy1Dto dummy1Dto = new Dummy1Dto();

        dummy1Dto.setId( dummy1.getId() );
        List<String> list = dummy1.getDescriptions();
        if ( list != null ) {
            dummy1Dto.setDescriptions( new ArrayList<String>( list ) );
        }

        return dummy1Dto;
    }

    @Override
    public Dummy1Dto toDtoLight(Dummy1 dummy1) {
        if ( dummy1 == null ) {
            return null;
        }

        Dummy1Dto dummy1Dto = new Dummy1Dto();

        dummy1Dto.setId( dummy1.getId() );

        return dummy1Dto;
    }
}
```

### Possible workarounds

The only workaround is to write BaseMapper#toDtoLight by hand.

### MapStruct Version

MapStruct 1.5.3

### PR DESCRIPTION
**Pull request by EvaristeGalois11**
Close #3119 

It's a simple fix that just passes along the qualifiers present on the BeanMapping annotation to aid the resolution of the subclass specific method.

I don't know if this is the right approach or you would prefer two specific new attributes on the SubclassMapping annotation. For my needs it is enough.

### CODE REVIEW

**Comment by EvaristeGalois11** [comment]
As pointed out in https://github.com/mapstruct/mapstruct/issues/3119#issuecomment-1364549035 using the qualifiers present in the BeanMapping annotation wasn't the right thing to do, so i expanded the SubclassMapping annotation with qualifiedBy and qualifiedByName.

File: processor/src/test/java/org/mapstruct/ap/test/subclassmapping/qualifier/SubclassQualifierMapperTest.java [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    }
+
+    @ProcessorTest
+    @WithClasses(ErroneousSubclassQualifiedByMapper.class)
+    @ExpectedCompilationOutcome(
+        value = CompilationResult.FAILED,
+        diagnostics = @Diagnostic(
+            type = ErroneousSubclassQualifiedByMapper.class,
+            kind = javax.tools.Diagnostic.Kind.ERROR,
+            line = 15,
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/subclassmapping/qualifier/SubclassQualifierMapperTest.java:233]: Ideally this should be the line with the `@SubclassMapping`, i.e. line 13

File: processor/src/test/java/org/mapstruct/ap/test/subclassmapping/qualifier/SubclassQualifierMapperTest.java [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    }
+
+    @ProcessorTest
+    @WithClasses(ErroneousSubclassQualifiedByNameMapper.class)
+    @ExpectedCompilationOutcome(
+        value = CompilationResult.FAILED,
+        diagnostics = @Diagnostic(
+            type = ErroneousSubclassQualifiedByNameMapper.class,
+            kind = javax.tools.Diagnostic.Kind.ERROR,
+            line = 15,
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/subclassmapping/qualifier/SubclassQualifierMapperTest.java:248]: Ideally this should be the line with the `@SubclassMapping`, i.e. line 13

**Review by filiphr — APPROVED** [review]
This looks great @EvaristeGalois11. Thanks a lot for the PR and sorry that it took so long to merge it.

I've left one comment, but I'll still merge the PR without it. I'll create an issue for the improvement of the error line location

**Comment by filiphr** [comment]
@EvaristeGalois11 I've created #3202 for the line location. In case you are interested in adding that, you have the first dibs on the issue :).

**Comment by EvaristeGalois11** [comment]
> This looks great @EvaristeGalois11. Thanks a lot for the PR and sorry that it took so long to merge it.

Thank you instead for accepting the PR, I had a really great time working on this project and i plan to contribute a bit more if i can in the future!

And no need to apologize about the wait, we all have personal stuff to take care :D

**Comment by filiphr** [comment]
>Thank you instead for accepting the PR, I had a really great time working on this project and i plan to contribute a bit more if i can in the future!

I am happy to hear that it was fun for you to work on this project. We appreciate any help we can get. So I won't say no for any help :)

**Pull request merged**